File: packages/semantic-cache/src/rerank.ts
/**
 * Built-in rerank factories for @betterdb/semantic-cache.
 */

/**
 * Tokenize: lowercase, split on whitespace, strip surrounding punctuation.
 * Deterministic and dependency-free.
 * IDF weighting would attach here at the token-weighting step.
 */
function tokenize(text: string): Set<string> {
  const out = new Set<string>();
  for (const raw of text.toLowerCase().split(/\s+/)) {
    const tok = raw.replace(/^[.,!?;:"'()\[\]{}<>]+|[.,!?;:"'()\[\]{}<>]+$/g, '');
    if (tok) out.add(tok);
  }
  return out;
}

/**
 * Built-in keyword-overlap reranker.
 *
 * Blends cosine similarity with word overlap and returns the index of the
 * best candidate.
 *
 * @param compare
 *   `"prompt"`  – overlap of the incoming query against each candidate's stored
 *                 prompt. Equivalence signal. Catches entity mismatches
 *                 (e.g. "weather in Paris" vs "weather in Berlin"). Default.
 *   `"response"` – overlap of the incoming query against each candidate's cached
 *                 response. Relevance signal.
 *
 * @param cosineWeight
 *   Weight on cosine similarity in [0, 1]. Overlap weight is `1 - cosineWeight`.
 *   Default: 0.7 (overlap 0.3).
 *
 * Candidate objects carry: `similarity` (cosine distance, lower = more similar),
 * `response` (string), and `prompt` (string, stored prompt).
 */
export function createKeywordOverlapRerank(options?: {
  compare?: 'prompt' | 'response';
  cosineWeight?: number;
}): (query: string, candidates: Array<{ response: string; similarity: number; prompt: string }>) => Promise<number> {
  const compare = options?.compare ?? 'prompt';
  const cosineWeight = options?.cosineWeight ?? 0.7;
  if (cosineWeight < 0 || cosineWeight > 1) {
    throw new Error('cosineWeight must be in [0, 1]');
  }
  const overlapWeight = 1.0 - cosineWeight;

  return async (query: string, candidates: Array<{ response: string; similarity: number; prompt: string }>): Promise<number> => {
    const queryTokens = tokenize(query);
    let bestIdx = 0;
    let bestScore = -Infinity;
    for (let i = 0; i < candidates.length; i++) {
      const text = candidates[i][compare] ?? '';
      const candTokens = tokenize(text);
      let overlap = 0;
      if (queryTokens.size > 0) {
        let intersection = 0;
        for (const t of queryTokens) {
          if (candTokens.has(t)) intersection++;
        }
        overlap = intersection / queryTokens.size;
      }
      const cosineSim = 1.0 - candidates[i].similarity;
      const score = cosineWeight * cosineSim + overlapWeight * overlap;
      if (score > bestScore) {
        bestScore = score;
        bestIdx = i;
      }
    }
    return bestIdx;
  };
}

File: packages/semantic-cache-py/betterdb_semantic_cache/rerank.py
"""Built-in rerank factories for betterdb-semantic-cache."""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Literal


def _tokenize(text: str) -> set[str]:
    """Lowercase, split on whitespace, strip surrounding punctuation.

    Deterministic and dependency-free.
    # IDF weighting would attach here at the token-weighting step.
    """
    out: set[str] = set()
    for raw in text.lower().split():
        tok = raw.strip(".,!?;:\"'()[]{}<>")
        if tok:
            out.add(tok)
    return out


def create_keyword_overlap_rerank(
    *,
    compare: Literal["prompt", "response"] = "prompt",
    cosine_weight: float = 0.7,
) -> Callable[[str, list[dict]], Awaitable[int]]:
    """Built-in keyword-overlap reranker.

    Blends cosine similarity with word overlap and returns the index of the
    best candidate.

    compare:
      "prompt"   - overlap of the incoming query against each candidate's stored
                   prompt. Equivalence signal. Catches entity mismatches
                   (e.g. "weather in Paris" vs "weather in Berlin"). Default.
      "response" - overlap of the incoming query against each candidate's cached
                   response. Relevance signal.

    cosine_weight: weight on cosine similarity in [0, 1]. Overlap weight is
                   (1 - cosine_weight). Default 0.7 (overlap 0.3).

    Candidate dicts are expected to carry: "similarity" (cosine distance, lower
    is more similar), "response" (str), and "prompt" (str, stored prompt).
    """
    if not 0.0 <= cosine_weight <= 1.0:
        raise ValueError("cosine_weight must be in [0, 1]")
    overlap_weight = 1.0 - cosine_weight

    async def rerank_fn(query: str, candidates: list[dict]) -> int:
        query_tokens = _tokenize(query)
        best_idx, best_score = 0, float("-inf")
        for i, cand in enumerate(candidates):
            text = str(cand.get(compare, "") or "")
            cand_tokens = _tokenize(text)
            if query_tokens:
                overlap = len(query_tokens & cand_tokens) / len(query_tokens)
            else:
                overlap = 0.0
            sim = float(cand.get("similarity", 1.0))
            cosine_sim = 1.0 - sim  # candidates carry cosine DISTANCE
            score = cosine_weight * cosine_sim + overlap_weight * overlap
            if score > best_score:
                best_score, best_idx = score, i
        return best_idx

    return rerank_fn
